tests/python_expression_dsl/test_evaluator.py

```diff
@@ -0,0 +1,138 @@
+from nose.tools import assert_raises, eq_
+from parameterized import parameterized
+
+from ...python_expression_dsl.evaluator import (
+    DSLEvaluationError,
+    DSLEvaluationVisitor,
+    DSLEvaluator,
+)
+from ...python_expression_dsl.parser import DSLParseError, DSLParser
+
+
+class Subject(object):
+    """Dummy object designed for testing DSLEvaluator."""
+
+    def __init__(self, attributes):
+        """Initialize a new instance of Subject.
+
+        :param attributes: List of attributes
+        :type attributes: List[str]
+        """
+        self._attributes = attributes
+
+    @property
+    def attributes(self):
+        """Return the list of attributes.
+
+        :return: List of attributes
+        :rtype: List
+        """
+        return self._attributes
+
+
+class TestDSLEvaluator(object):
+    @parameterized.expand(
+        [
+            ("incorrect_expression", "?", None, None, DSLParseError),
+
+            ("numeric_literal", "9", 9),
+            ("numeric_float_literal", "9.5", 9.5),
+
+            ("unknown_identifier", "foo", None, None, DSLEvaluationError),
+            ("known_identifier", "foo", 9, {"foo": 9}),
+            ("unknown_nested_identifier", "foo.bar", None, {"foo": 9}, DSLEvaluationError),
+            ("known_nested_identifier", "foo.bar", 9, {"foo": {"bar": 9}}),
+            (
+                "identifier_pointing_to_the_object",
+                "'eresources' in subject.attributes",
+                True,
+                {"subject": Subject(["eresources"])},
+            ),
```

Comment:
Using a DSL expression you can only get access to only what's inside the context which can be either a dictionary or a Python object. The test case above emulates the scenario I'm planning to implement in SAML enhancements PR. In that case the DSL engine will have access only to `api.saml.metadata.model.SAMLSubject`, all of its nested properties and some built-in Python functions:
```python
    BUILTIN_FUNCTIONS = {
        "abs": abs,
        "all": all,
        "any": any,
        "len": len,
        "max": max,
        "min": min,
    }
```

Reply:
My concern is with the nested properties and method calls. If you put a SQLAlchemy data model object into the context, then executing `obj.delete()` will delete that object from the database when the request completes. If nested properties also work, executing something like `obj.patron.delete()` wil delete some _other_ object from the database.

Maybe there is no solution apart from code review of anything that uses this system to make sure it doesn't put unsafe objects into the context. I was hoping we could avoid that by limiting the DSL's expressive capability to something closer to XPath.

Reply:
My main intent was to make the DSL engine almost as much powerful as Python so future users could write filter expression of any complexity but I do see your point. We can try to run DSL expression in a [sandbox](https://stackoverflow.com/questions/3068139/how-can-i-sandbox-python-in-pure-python) by, for example, executing them in a separate `exec` context.

Reply:
The other thing that we can do is to make a deep copy of context objects by serializing and deserializing them before putting into the context - it will allow to unbind them from CM's execution context, SQLAlchemy session, etc.